File: Graph/BFS/1284. Minimum Number of Flips to Convert Binary Matrix to Zero Matrix.cpp
class Solution {
public:
    int minFlips(vector<vector<int>>& mat) {
        const int m = mat.size();
        const int n = mat[0].size();
        const vector<int> dirs{0, 1, 0, -1,  0, 0};

        auto flip = [&](int s, int x, int y) {
            for (int i = 0; i < 5; i++) {
                const int tx = x + dirs[i];
                const int ty = y + dirs[i + 1];
                if (tx < 0 || tx >= n || ty < 0 || ty >= m) {
                    continue;
                }
                s ^= (1 << (ty * n + tx));
            }
            return s;
        };

        int steps = 0;
        queue<int> q;
        vector<int> seen (1 << (m * n));
        int start = 0;
        for (int y = 0; y < m; y++) {
            for (int x = 0; x < n; x++) {
                start |= (mat[y][x] << (y * n + x));
            }
        }
        q.push(start);
        seen[start] = 1;

        while (!q.empty()) {
            int sz = q.size();
            while (sz--) {
                int s = q.front();
                if (s == 0) return steps;
                q.pop();
                for (int y = 0; y < m; y++) {
                    for (int x = 0; x < n; x++) {
                        int t = flip(s, x, y);
                        if (seen[t]) {
                            continue;
                        }
                        q.push(t);
                        seen[t] = 1;
                    }
                }
            }
            steps++;
        }
        return -1;
    }
};

File: Graph/BFS/1284. Minimum Number of Flips to Convert Binary Matrix to Zero Matrix.java
class Solution {
    private static final int[][] directions = new int[][] {{ -1, 0}, {1, 0}, {0, -1}, {0, 1}, {0, 0}};
    public int minFlips(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;

        int[] visited = new int[1 << (m * n)];
        Queue<Integer> queue = new ArrayDeque<>();
        int start = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                start |= (mat[i][j] << i * n + j);
            }
        }
        queue.offer(start);
        visited[start] = 1;
        int step = 0;
        while (!queue.isEmpty()) {
            int sz = queue.size();
            for (int i = 0; i < sz; i++) {
                int state = queue.poll();
                if (state == 0) {
                    return step;
                }
                for (int x = 0; x < m; x++) {
                    for (int y = 0; y < n; y++) {
                        int ns = flip(state, x, y, m, n);
                        if (visited[ns] == 0) {
                            queue.offer(ns);
                            visited[ns] = 1;
                        }
                    }
                }
            }
            step++;
        }

        return -1;
    }

    private int flip (int state, int x, int y, int m, int n) {
        for (int[] dir : directions) {
            int nx = x + dir[0];
            int ny = y + dir[1];
            if (nx < 0 || nx >= m || ny < 0 || ny >= n) {
                continue;
            }
            state ^= (1 << (nx * n + ny));
        }

        return state;
    }
}